### ISSUE TITLE
Loader needs to use different check for empty directories in OS X (for jar)

### ISSUE THREAD

**Issue by flaw600** [ATLauncher/ATLauncher#198]
One of ATLauncher's principles is portability. That principle requires the folder the jar to be put into an empty directory. However, the check for empty directories is insufficient. Currently Loader checks to make sure that there is only 1 file in the directory at time of initial launch. That is, that the jar is the only file in the folder. If it is, the folder is 'empty'. However, by that definition, no folder in OS X is ever truly empty. This is because OS X utilizes a file called .DS_Store. This is a system file that is put inside every folder (and subfolder) at time of creation and cannot be permanently erased unless you do some configuration with Spotlight. The Loader's check for an empty folder picks up that file and thus declares the folder to be "non-empty". Thus, a new check is needed for OS X where it checks for # of files and if (files >1) /and/ the file picked up isn't .DS_Store, /then/ the folder is non-empty. Otherwise, the warning may panic the user needlessly.

**Comment by dgelessus** [comment]
Or ignore all files starting with a `.`, which are by convention "hidden" files on Unix-based systems.

**Comment by flaw600** [comment]
@dgelessus does this sort of problem happen in Linux systems? Genuine question. .DS_Store is unique to OS X from what I understand. But if you want to ignore all ‘.’ files that would work as well. Actually, come to think of it, it’s cleaner too. Thanks for the suggestion.

**Comment by flaw600** [comment]
@dgelessus do you like this implementation better?

**Comment by dgelessus** [comment]
I don't actively use any Linuxes, so I don't know if there are any common programs that automatically create dot files. What I know though is that `.DS_Store` is by far not the only dot file that the Finder creates. Most importantly files in the trash are stored in a `.Trashes` subfolder of their original location - this could be especially confusing when a user trashes all files from the ATL folder without emptying the trash. If you use the Time Machine backup feature, there are also `.com.apple.timemachine.supported` files. There are also a few other dot files like `.localized` or `.background`.

Long story short, it should be safe to ignore dot files. The launcher doesn't create any itself, so they are extremely unlikely to cause any problems.

**Comment by flaw600** [comment]
Ah, okay. I made a new method isSystemFile to check if a given file returned by .list is a 'dot' (system) file. I use it in the (file >1) section to check if it is /not/ a system file and only then to display the dialog. This was done in the last 2 commits in my fork. Hopefully they are referenced here in this thread. I see it but idk if others can. If you could take a look at it and see if this implementation is done correctly (aka coding style), that'd be great. Thanks!

**Comment by dgelessus** [comment]
Or even better - use the standard `isHidden()` method of `File`s. That also covers anything the operating system considers hidden, such as Windows Explorer's `desktop.ini`.

Can't help you much with checking/merging your changes though, I don't do any actual development or work on the launcher source code. I'm also not quite sure which branch you should send your pull request to - ATLauncher/master is probably the wrong one, but in ATLauncher/3.3.0.0 there's a lot of internal restructuring going on. ping @RyanTheAllmighty

**Comment by flaw600** [comment]
Oh yeah, isHidden is a thing. I’ll do that now. Thanks!

**Comment by s0cks** [comment]
Please use .listFiles with a file filter - filtering out the hidden files please its cleaner.....And more efficient

**Comment by flaw600** [comment]
I did. Please see the commit linked (it should new visible) to the issue.

**Comment by s0cks** [comment]
I dont see the use of file filter just a static var and iteration

**Comment by flaw600** [comment]
Please explain what you mean. I get the file list and check for hidden files and if there are the dialog doesn't display.

**Comment by s0cks** [comment]
File#listFiles accepts an instance of FileFilter please use that function

**Comment by flaw600** [comment]
I fail to see how that's cleaner. In fact, it results in more code to be written since you then have to write the file filter class and accept method. It is neither more efficient nor cleaner

**Comment by s0cks** [comment]
Its cleaner because you aren't needing non-local variables, its more efficient because your looping twice one for the listFiles call and one for the sorting. Using FileFilter loops once, plus yours isn't thread safe mine would be

**Comment by RyanTheAllmighty** [comment]
For one it's better since you can filter out the hidden files in the filter, which provides a better way to count the number of results, since in the case of what you've written as soon as it comes across a hidden file it returns true, but in the case of there being hidden and regular files and the hidden pops up in the loop before the regular, it returns true and doesn't count any regular files

**Comment by s0cks** [comment]
Ryan also has a good point

**Comment by flaw600** [comment]
Ryan: yeah I just realized that 5 minutes before I read this. Thanks for mentioning it though.

s0cks: ooc, how does mine loop twice? What part are you referring to? I don't see it. As for thread safe, I'll take your word for it. Also, the non local variables are done that way cause the compiler complained if they weren't that way.

At any rate, I was supposed to be asleep by now so I'll take a look at it tomorrow. Thanks for the suggestions!

**Comment by s0cks** [comment]
List files without a file filter still loops through the available files, your doing a second iteration to  sort them, if you use a filter then your eliminating the second loop because your sorting intern

**Comment by flaw600** [comment]
I suppose listFiles(FileFilter) is marginally more efficient. Instead of two loops like I was doing, we have a loop and an if statement. And I suppose it is cleaner, albeit slightly. At any rate, please look at the latest commit.

**Comment by flaw600** [comment]
@RyanTheAllmighty does the current implementation work better?

**Comment by s0cks** [comment]
Something I've been meaning to do, and before Ryan or someone implements this on the 3.3.0.0 branch, Can you please change the anonymous class into a regular class as anonymous classes affect performance, as they are turned into non-static inner-classes which can cause memory leaks through needing an implicit reference to the outer-class which is in turn an object sitting in memory not able to be accessed.

**Comment by s0cks** [comment]
I'm going to be working on converting inner and anonymous classes into regular/static-inner classes soon

**Comment by flaw600** [comment]
As I said in #199 I’m more interested in how big of a memory leak these types of classes actually cause. If it’s a trivial amount then is it really worth the effort to do go through all the classes? At any rate, I’ve done as you asked. pinging @RyanTheAllmighty to confirm that this conforms to expectations

**Comment by RyanTheAllmighty** [comment]
Looks good. Send in the PR to 3.3.0.0 branch and I'll pop it in
